feat: Ученики мага при ХАЙПОПе

PhantornRU opened the pull request:
Авоут пройден и одобрен:
https://discord.com/channels/617003227182792704/755125334097133628/1020299400708366356
Пройденные голосования по школам магов:
https://discord.com/channels/617003227182792704/1005821829748895875/1020659750167384105
https://discord.com/channels/617003227182792704/755125334097133628/1024583287236202526

## Fixes
https://github.com/ss220-space/Paradise/pull/1569
https://github.com/ss220-space/Paradise/pull/1584
https://github.com/ss220-space/Paradise/pull/1586

## What Does This PR Do
![image](https://user-images.githubusercontent.com/41479614/191004136-4a3cf091-8d85-491f-adfc-ab057f2aa0d5.png)

- Добавление учеников к гейм-моду мага при ХАЙПОПЕ.
- Ученики появляются раундстартово и получают цель на защиту учителя. Если учитель меняет имя, имя меняется на цели.
- Ученики по прежнему не могут трогать учительские вещи.
- В антаг-панеле, персонажу можно выдать роль мага или роль ученика. Только что-то одно.
- При выдаче ролей ученика, он подстраивается под свой лодаут.
- Добавление отдельного фиолетового магического учебника для ученика мага, позволяющего выбирать наборы заклинаний.
- Внесение учеников в разные "мероприятия".
- Небольшой рефактор контрактов для более удобного наследования и переделывания.
- Контракт перенесен в отдельный файл, т.к. перестал быть только "артефактом".
- Расширение магического автомата, чтобы ученики тоже могли переодеться.  Добавление магических вещей, ранее не участвующих ни в чем и просто висящие в билде без дела.
- Одежда в магическом автомате с 1 штуки повышена до 5 штук.
- Фикс ЭПИЧЕСКИХ пятюнь которые вызывали некорректные ЭПИЧЕСКИЕ взрывы. Дополнение их учениками.

После 50 человек, каждые 25 человек, вместе с магом будут также появляться его ученики (выбираются из лобби, как и маг).
Напомню, ученики имеют 1 свиток телепорта и 1 из 4-х наборов заклинаний.

### Пропорции:
<49 - 1 маг, 0 учеников
50-74 - 1 маг, 1 ученик
75-99 - 1 маг, 2 ученика
100< - 1 маг, 3 ученика

## Why It's Good For The Game
- Теперь погибший маг на хайпопе от толпы асистух/СБух - не означает эксту
- Люди теперь больше приспособлены к магам, так как больше за них играют.

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif of your feature if you want -->
![image](https://user-images.githubusercontent.com/41479614/190857280-c8c46fee-d219-445a-b533-ae30a2fc647e.png)
![image](https://user-images.githubusercontent.com/41479614/191004256-03fb00e3-f515-4469-87ac-b1824cc7b3f9.png)
![image](https://user-images.githubusercontent.com/41479614/190857378-001b3535-d833-4b08-8935-45d07b20cb07.png)
![image](https://user-images.githubusercontent.com/41479614/191004098-ac95610a-1cf2-4fe7-a853-6a0d879cf2d3.png)

PhantornRU committed:
Добавлен ученик волшебника при ХАЙПОПЕ

PhantornRU committed:
Фикс некорректного взрыва при выдаче пятюней магами друг другу.

Ученики тоже могут дать пятюню. С вредом для себя.

larentoun commented on code/datums/mind.dm:
```diff
@@ -236,9 +236,13 @@
 		if(objectives.len==0)
 			. += "<br>Objectives are empty! <a href='?src=[UID()];wizard=autoobjectives'>Randomize!</a>"
+	else if(src in SSticker.mode.apprentices)
+		. += "<b><font color='red'>WIZARD APPRENTICE</font></b>|<a href='?src=[UID()];wizard=clear'>no</a>"
```
Я так понимаю, span class не работает?

PhantornRU replied:
Он выдает не совсем то что хотелось бы видеть + это оригинальный код. Оставлено во избежании изменений в шрифтах и жалоб.

PhantornRU replied:
Заменен на спан

larentoun commented on code/datums/mind.dm:
```diff
@@ -236,9 +236,13 @@
+	else if(src in SSticker.mode.apprentices)
+		. += "<b><font color='red'>WIZARD APPRENTICE</font></b>|<a href='?src=[UID()];wizard=clear'>no</a>"
+		. += "<br><a href='?src=[UID()];wizard=lair'>To lair</a>, <a href='?src=[UID()];common=undress'>undress</a>, <a href='?src=[UID()];wizard=dressup'>dress up</a>, <a href='?src=[UID()];wizard=name'>let choose name</a>."
+		if(objectives.len==0)
```
Ну, можно было бы проверять на наличие длины, а не ее равность нулю

PhantornRU replied:
нужно провести рефактор всех упоминаний. Я не могу провести рефактор в данном ПРе для всех таких проверок, необходим отдельный ПР.
Иначе меня опять будут ругать что всё в одном смешиваю.

larentoun replied:
Не пинают за рефакторы

PhantornRU replied:
И то верно. Тогда Впишу спан классы и проведу небольшие рефакторы.

PhantornRU replied:
Внесена проверка от длины

larentoun commented on code/datums/mind.dm:
```diff
@@ -887,8 +891,17 @@
+					current.spellremove(current)
+					current.faction = list("Station")
+					SSticker.mode.update_wiz_icons_removed(src)
+					to_chat(current, "<span class='warning'><FONT size = 3><B>You have been brainwashed! You are no longer a apprentice wizard!</B></FONT></span>")
```
Почему бы не использовать danger вместо warning? Насколько я помню, он как раз увеличивает размер текста

PhantornRU replied:
То что было изначально в коде, то и оставил, дабы не было возмущений о "НИЕЕЕТ, ТЫ ПЕРЕДЕЛАЛ НЕ ПО ПРЕДЛОЖКЕ" и мгновенного отката.

larentoun replied:
Не стоит добавлять фичи, да. Но код-стандарты никто не отменял

PhantornRU replied:
И то верно. Тогда Впишу спан классы.

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+	if(teacher.stat || teacher.restrained())
+		return
+	if(!istype(teacher, /mob/living/carbon/human))
```
ishuman()

PhantornRU replied:
Поправлено

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+/obj/item/contract/apprentice/Topic(href, href_list)
+	..()
+	var/mob/living/carbon/human/teacher = usr
```
А проверка на то, что он хуман не нужна, точно?

PhantornRU replied:
А как ты заюзаешь предмет в руке, не будучи хуманом?

larentoun replied:
Возможно педальская магия. В любом случае, желательно сначала делать на тип перед тем, как приравнивать к этому типу

PhantornRU replied:
![image](https://user-images.githubusercontent.com/41479614/191005420-9091ef32-51f6-4421-8dd8-2e68e032e87e.png)

Так я так и так проверку на потом выношу. Её же достаточно всегда бывает.

larentoun replied:
А, ну, я бы тогда проверку хумана просто выше бы переставила

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+	if(!istype(teacher, /mob/living/carbon/human))
+		return 1
+
+	if(loc == teacher || (in_range(src, teacher) && istype(loc, /turf)))
```
isturf()

PhantornRU replied:
Исправлено

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+		teacher.set_machine(src)
+		if(href_list["school"])
+			if(used)
+				to_chat(teacher, "You already used this contract!")
```
Span class notice/warning?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+				to_chat(teacher, "You already used this contract!")
+				return
+			used = 1
+			to_chat(teacher, "apprentice waiting...")
```
Span class notice?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			to_chat(teacher, "apprentice waiting...")
+			var/image/source = image('icons/obj/cardboard_cutout.dmi', "cutout_wizard")
+			var/list/candidates = SSghost_spawns.poll_candidates("Do you want to play as the wizard apprentice of [teacher.real_name]?", ROLE_WIZARD, TRUE, source = source)
+			if(candidates.len)
```
Все еще рекомендую length(), на случай, если candidates будет null

PhantornRU replied:
заменил

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			if(candidates.len)
+				var/mob/C = pick(candidates)
+				new /obj/effect/particle_effect/smoke(teacher.loc)
+				var/mob/living/carbon/human/apprentice = new/mob/living/carbon/human(teacher.loc)
```
Может лучше в турфе тичера, а не в локе? А если он в мехе, шкафу, роде и тд?

PhantornRU replied:
Свиток юзается в Wizard's Den. Там есть только 1 шкаф. Но даже в шкафе он корректно ставит.
В роде нельзя заюзать свиток.
Магу нельзя в мехов.

+ это оригинальный код, не трогал его.

larentoun replied:
Всё же лучше добавить защиту от дурака, а то появятся магические мехи и всё, капут

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+				new /obj/effect/particle_effect/smoke(teacher.loc)
+				var/mob/living/carbon/human/apprentice = new/mob/living/carbon/human(teacher.loc)
+				apprentice.key = C.key
+				to_chat(apprentice, "<B>You are the [teacher.real_name]'s apprentice! You are bound by magic contract to follow [teacher.p_their()] orders and help [teacher.p_them()] in accomplishing their goals.")
```
Span class, тот же danger?

PhantornRU replied:
Нотис вставил.

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+				apprentice.real_name = newname
+				apprentice.name = newname
+				var/datum/objective/protect/new_objective = new /datum/objective/protect
+				new_objective.owner = apprentice:mind
```
Уверен ли ты, что хочешь использовать : вместо .

PhantornRU replied:
Это оригинальный код. Я вообще без понятия что оно делает. В другом месте я это выпилил. Тут код оставлен.

larentoun replied:
Ошибки с `:` буду замечены только в рантайме. Следовало изучить этот вопрос. Советую сменить на `.` для большей надежности.

PhantornRU replied:
исправил

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+				apprentice.faction = list("wizard")
+			else
+				used = 0
+				to_chat(teacher, "Unable to reach your apprentice! You can either attack the spellbook with the contract to refund your points, or wait and try again later.")
```
Span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+	if(apprentice.stat || apprentice.restrained())
+		return
+	if(!istype(apprentice, /mob/living/carbon/human))
```
ishuman()

PhantornRU replied:
Исправлено

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+/obj/item/contract/apprentice_choose_book/Topic(href, href_list)
+	..()
+	var/mob/living/carbon/human/apprentice = usr
```
Точно уверен, что всегда будет хуманом?

PhantornRU replied:
Ты не сможешь использовать книжку в другом облике

larentoun replied:
Руки есть не только у хуманов

PhantornRU replied:
Для тех у кого ручки - всеравно нельзя пользоваться.
![image](https://user-images.githubusercontent.com/41479614/191012749-76a05bf0-9e8f-4fbe-ab6e-959d7f6bf815.png)

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+	if(!istype(apprentice, /mob/living/carbon/human))
+		return 1
+
+	if(loc == apprentice || (in_range(src, apprentice) && istype(loc, /turf)))
```
isturf()

PhantornRU replied:
Исправлено

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+		apprentice.set_machine(src)
+		if(href_list["school"])
+			if(used)
+				to_chat(apprentice, "Учебник уже был вами изучен!")
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+/////////Choose Spells Pack//////////
+
+/obj/item/contract/proc/school_href_choose(href_list, var/mob/living/carbon/human/teacher, var/mob/living/carbon/human/apprentice)
```
можно без var/

PhantornRU replied:
действительно. Поправил.

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/projectile/magic_missile(null))
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/click/fireball(null))
+			if (teacher)
+				to_chat(teacher, "<B>Your service has not gone unrewarded, however. Studying under [teacher.real_name], you have learned powerful, destructive spells. You are able to cast magic missile and fireball.")
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/area_teleport/teleport(null))
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/ethereal_jaunt(null))
+			if (teacher)
+				to_chat(teacher, "<B>Your service has not gone unrewarded, however. Studying under [teacher.real_name], you have learned reality bending mobility spells. You are able to cast teleport and ethereal jaunt.")
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/forcewall(null))
+			apprentice.equip_to_slot_or_del(new /obj/item/gun/magic/staff/healing(apprentice), slot_r_hand)
+			if (teacher)
+				to_chat(teacher, "<B>Your service has not gone unrewarded, however. Studying under [teacher.real_name], you have learned livesaving survival spells. You are able to cast charge and forcewall.")
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/aoe_turf/knock(null))
+			apprentice.mind.AddSpell(new /obj/effect/proc_holder/spell/targeted/click/mind_transfer(null))
+			if (teacher)
+				to_chat(teacher, "<B>Your service has not gone unrewarded, however. Studying under [teacher.real_name], you have learned stealthy, robeless spells. You are able to cast knock and mindswap.")
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+
+///Сообщение выдаваемое при использовании использованных контрактов
+/obj/item/contract/proc/used_contract()
+	return "<B>You have already summoned your apprentice.</B><BR>"
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+	return "<B>You have already summoned your apprentice.</B><BR>"
+
+/obj/item/contract/apprentice_choose_book/used_contract()
+	return "<B>Вами уже был изучен учебник.</B><BR>"
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+	dat += "<I>Если учитель не настроен на разговор - ничего страшного! В данном учебнике приведено краткое описание возможных путей.</I><BR>"
+
+	dat += "<B>Какую школу магии вы хотели бы изучать?:</B><BR>"
+	return dat
```
span class?

larentoun commented on code/game/gamemodes/wizard/apprentice.dm:
```diff
@@ -0,0 +1,179 @@
+	dat += "<I>If you are unable to establish contact with your apprentice, you can feed the contract back to the spellbook to refund your points.</I><BR>"
+
+	dat += "<B>Which school of magic is your apprentice studying?:</B><BR>"
+	return dat
```
span class?

larentoun commented on code/game/gamemodes/wizard/wizard.dm:
```diff
@@ -33,15 +36,36 @@
+	possible_wizards.Remove(wizard)
+	if(playerC >= required_min_players_for_apprentice)
+		for(var/i in 0 to ((playerC - required_min_players_for_apprentice) / required_num_players_for_apprentice))
+			if(possible_wizards.len==0)
```
Можно просто проверку на длину, а не сравнивать с нулем

PhantornRU replied:
Исправил

larentoun commented on code/game/gamemodes/wizard/wizard.dm:
```diff
@@ -166,6 +219,54 @@
+	qdel(wizard_mob.r_store)
+	qdel(wizard_mob.l_store)
+
+	if(isplasmaman(wizard_mob))
```
На воксов проверка не нужна?

PhantornRU replied:
Внёс для воксов

larentoun commented on code/game/gamemodes/wizard/wizard.dm:
```diff
@@ -166,6 +219,54 @@
+
+
+
+	to_chat(wizard_mob, "Вы найдёте набор из доступных закинаний в вашем магическом учебнике.")
```
span calss

larentoun commented on data/mode.txt:
```diff
@@ -1 +1 @@
-extended
+secret
```
:XMM:

PhantornRU replied:
блят откуда, я ж убирал >:C

PhantornRU committed:
Логические магические правки

PhantornRU committed:
Добавление школ колдунств по предложке

https://discord.com/channels/617003227182792704/1005821829748895875/1020659750167384105

PhantornRU committed:
Фикс священного огня (он не должен был наносить урон магу)

ПРы ФИКСОВ огнеупорности с ОФФов:
https://github.com/ParadiseSS13/Paradise/pull/15817
https://github.com/ParadiseSS13/Paradise/pull/15815

PhantornRU committed:
Фикс надевания декоративных балахонов разных школ, фикс не выдачи трейтов

PhantornRU committed:
Рефакторы, спан классы

PhantornRU commented:
@larentoun внёс спан классы, отрефакторил всё-таки обжективы.

А еще красивый пикч добавил :)))

PhantornRU committed:
Откат спан класса, не отображающегося в антаг-панеле админов.

larentoun commented on code/datums/mind.dm:
```diff
@@ -208,7 +208,7 @@
 
 		. += " <a href='?src=[UID()];revolution=reequip'>Reequip</a> (gives security HUD and spray can)."
-		if(objectives.len==0)
+		if(!length(objectives.len))
```
Аа... Эээ... А зачем дважды длина?)
Хватило бы просто length(objectives), так как objectives задается ведь как лист

Ну и далее тоже пофиксить этот прикол, на каждый оставлять не буду коммент

PhantornRU replied:
Действительно, очепятался, заменил

PhantornRU replied:
Done

larentoun reviewed (changes requested):
Ну, вроде последнее

PhantornRU committed:
Вынесение проверки на гуманоидность мага, исправление ошибки с длиной

larentoun reviewed (approved):
Ну, на мой скромный взгляд, вроде нормально
